controller/gui/client.js

```diff
@@ -65,3 +61,57 @@ customElements.define(
   },
   { extends: 'input' }
 )
+
+/* Form web component preventing more than one submission.
+ *
+ * Disable submit input inside the form after the first submission.
+ */
+customElements.define(
+  'disable-after-submit',
+  class extends HTMLFormElement {
+    constructor() {
+      super()
+
+      const form = this
+      const button = form.querySelector('input[type=submit]')
+
+      const buttonParent = installParent(button, document.createElement('span'))
+      buttonParent.style = `
+        position: relative;
+        height: fit-content;
+      `
+
+      const spinnerParent = document.createElement('div')
+      spinnerParent.style = `
+        display: flex;
+        position: absolute;
+        top: 50%;
+        left: 50%;
+        transform: translate(-50%, -50%);
+      `
+
+      const spinner = document.createElement('span')
+      spinner.className = 'd-Spinner'
+      spinnerParent.appendChild(spinner)
+
+      form.addEventListener('submit', function() {
+        button.disabled = true
+        button.style.color = 'transparent'
+        button.className += ' d-Button--Disabled'
+        buttonParent.appendChild(spinnerParent)
+      })
+    }
+  },
+  { extends: 'form' }
+)
+
+/* Place given node under a new parent node.
+ *
+ * Useful to extend nodes that can not have childreen in web components, for
+ * ex. inputs.
+ */
+function installParent(node, newParent) {
```

Comment:
What do you think of the name `wrap` instead? As I understand this function basically wraps `node` with `newParent`. It took me some effort to understand what `installParent` does; naming it `wrap` would make it more intuitive.

Reply:
https://github.com/dividat/playos/pull/83/commits/43fd0021f1daf2c46e58d08fe4c1d3edbc026c32